htlcswitch/iterator.go

```diff
@@ -194,32 +239,141 @@ func (p *OnionProcessor) DecodeHopIterator(r io.Reader, rHash []byte) (HopIterat
 		}
 	}
 
-	return &sphinxHopIterator{
-		nextPacket:      sphinxPacket.NextPacket,
-		processedPacket: sphinxPacket,
-	}, lnwire.CodeNone
+	return makeSphinxHopIterator(onionPkt, sphinxPacket), lnwire.CodeNone
 }
 
-// ExtractErrorEncrypter takes an io.Reader which should contain the onion
-// packet as original received by a forwarding node and creates an
-// ErrorEncrypter instance using the derived shared secret. In the case that en
-// error occurs, a lnwire failure code detailing the parsing failure will be
-// returned.
-func (p *OnionProcessor) ExtractErrorEncrypter(r io.Reader) (ErrorEncrypter, lnwire.FailCode) {
+// DecodeHopIterators performs batched decoding and validation of incoming
+// sphinx packets. For the same `id`, this method will return the same iterators
+// and failcodes upon subsequent invocations.
+//
+// NOTE: In order for the responses to be valid, the caller must guarantee that
+// the presented readers and rhashes *NEVER* deviate across invocations for the
+// same id.
+func (p *OnionProcessor) DecodeHopIterators(id []byte, rs []io.Reader,
```

Comment:
API comment: rather than passing a series of slices of diff types, we could instead pass a slice of a single struct. Something like: 
```go
type SerializedHopIterator struct {
    IteratorBytes io.Reader
    RHash [32]byte
    IncomingCLTV
}
```

Same goes for the response value:
```go
type ProcessedHopIterator (prob a better name to use) struct {
    Iterator HopIterator
    FailCode lnwire.FailCode
}
```

Reply:
Fixed.